## Creating the game_logs for the basic statistics for every player based on retrosheet data.

I want to create a game_log sheet for every player in my data set. I realized this will help accelerate my backtesting. 

<br><br> The reason I am creating this file is that once I run the script once. I will just need to call the specific games to get the required information instead of generating it every time and not saving the results

<br> https://www.baseball-reference.com/players/gl.fcgi?id=bautijo02&t=b&year=2014  <br><br> this is the ideal format except the CSV file created will have all the players. So far it works from 2005-2018. File might be hundreds of MB but that is OK.


In [1]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

d = pd.read_csv('../2005-2018_games.csv')
d.rename(columns={'unknown':'double_header_flag'}, inplace=True)
d['ab_flag'] = d.ab_flag.map({'F':0,'T':1})
d.sh_flag = d.sh_flag.map({'F':0,'T':1})
d.sf_flag = d.sf_flag.map({'F':0,'T':1})

In [2]:
df = d.copy()

In [3]:
#find me all the Plate Apperances as described by MLB rules
AB_only = df[df.ab_flag == 1]

event_flags = [14,15,16,17] #14 is BB 15 is IBB 16 is HBP 17 is interference
BB_IBB_HBP_INT = df[df.event_type.isin(event_flags)]

sac_hits = df[df.sh_flag ==1]

sac_fly  = df[df.sf_flag ==1]

#combine all the dataFrames into 1.
new_df = AB_only.append(BB_IBB_HBP_INT)
new_df = new_df.append(sac_hits)
new_df = new_df.append(sac_fly)

df = new_df.copy()

In [4]:
hit_flags = [20,21,22,23]     #20 = single, 21= double, 22 = triple, 23 = homerun
df.loc[df.event_type.isin(hit_flags), 'hit_flag'] = 1
df.loc[df.event_type.isin(hit_flags) == False, 'hit_flag'] = 0

#event_type 14 is BB, 15 is IBB, 16 is HBP, 17 is interference
df.loc[df.event_type==14, 'BB'] = 1
df.loc[df.event_type!=14, 'BB'] = 0

df.loc[df.event_type==15, 'IBB'] = 1
df.loc[df.event_type!=15, 'IBB'] = 0

df.loc[df.event_type==16, 'HBP'] = 1
df.loc[df.event_type!=16, 'HBP'] = 0

df.loc[df.event_type==17, 'ITF'] = 1
df.loc[df.event_type!=17, 'ITF'] = 0

df.loc[df.event_type==20, '1B'] = 1
df.loc[df.event_type!=20, '1B'] = 0

df.loc[df.event_type==21, '2B'] = 1
df.loc[df.event_type!=21, '2B'] = 0

df.loc[df.event_type==22, '3B'] = 1
df.loc[df.event_type!=22, '3B'] = 0

df.loc[df.event_type==23, 'HR'] = 1
df.loc[df.event_type!=23, 'HR'] = 0

In [5]:
data_grouped = df.groupby(['game_id','year','month','day_','res_batter'])

#probably need to change game_id to double_header_flag that way but for now keep since you can just copy paste game_id on google and find the right game

In [6]:
# now that everything is seperated by game_id, 'year', 'month', 'day' and 'batter name' aggregate the stats
batting_summary = data_grouped.agg({'hit_flag':[np.mean,np.sum, np.size],'ab_flag':[np.sum],'BB':np.sum,'IBB':np.sum,'HBP':np.sum,'ITF':np.sum,'1B':np.sum,'2B':np.sum,'3B':np.sum,'HR':np.sum,'sh_flag':np.sum,'sf_flag':np.sum})

In [7]:
#Just some cleaning
batting_hits_pa = batting_summary.hit_flag
batting_hits_pa.rename(columns={'mean':'H/PA','sum':'H','size':'PA'}, inplace=True)

batting_ab = batting_summary.ab_flag
batting_ab.rename(columns={'sum':'AB'}, inplace=True)

batting_BB  = batting_summary.BB
batting_BB.rename(columns={'sum':'BB'}, inplace=True)

batting_IBB = batting_summary.IBB
batting_IBB.rename(columns={'sum':'IBB'}, inplace=True)

batting_HBP = batting_summary.HBP
batting_HBP.rename(columns={'sum':'HBP'}, inplace=True)

batting_ITF = batting_summary.ITF
batting_ITF.rename(columns={'sum':'ITF'}, inplace=True)

batting_1B = batting_summary['1B']
batting_1B.rename(columns={'sum':'1B'}, inplace=True)

batting_2B = batting_summary['2B']
batting_2B.rename(columns={'sum':'2B'}, inplace=True)

batting_3B = batting_summary['3B']
batting_3B.rename(columns={'sum':'3B'}, inplace=True)

batting_HR = batting_summary.HR
batting_HR.rename(columns={'sum':'HR'}, inplace=True)

batting_SH = batting_summary.sh_flag
batting_SH.rename(columns={'sum':'SH'}, inplace=True)

batting_SF = batting_summary.sf_flag
batting_SF.rename(columns={'sum':'SF'}, inplace=True)


#now combine all the smaller dataframes
batting_summary = pd.concat([batting_hits_pa,batting_ab,batting_BB,batting_IBB,batting_HBP,batting_ITF,batting_1B,batting_2B,batting_3B,batting_HR,batting_SH,batting_SF], axis =1)
batting_summary.sort_values(by=['res_batter','year','month','day_'],inplace=True)

#calculate BA and OBP: BA = H / AB, OBP = (H + BB + IBB + HBP)/ (AB + BB + IBB + HBP + SF )
batting_summary['BA'] = batting_summary.groupby(['res_batter','year'])['H'].transform(pd.Series.cumsum) / batting_summary.groupby(['res_batter','year'])['AB'].transform(pd.Series.cumsum)
batting_summary['OBP'] = (batting_summary.groupby(['res_batter','year'])['H'].transform(pd.Series.cumsum) + batting_summary.groupby(['res_batter','year'])['BB'].transform(pd.Series.cumsum) + batting_summary.groupby(['res_batter','year'])['IBB'].transform(pd.Series.cumsum) + batting_summary.groupby(['res_batter','year'])['HBP'].transform(pd.Series.cumsum) ) / (batting_summary.groupby(['res_batter','year'])['AB'].transform(pd.Series.cumsum) + batting_summary.groupby(['res_batter','year'])['BB'].transform(pd.Series.cumsum) + batting_summary.groupby(['res_batter','year'])['IBB'].transform(pd.Series.cumsum) + batting_summary.groupby(['res_batter','year'])['HBP'].transform(pd.Series.cumsum) + batting_summary.groupby(['res_batter','year'])['SF'].transform(pd.Series.cumsum))

batting_summary.reset_index(inplace=True)

In [8]:
batting_summary.query('res_batter=="bautj002" and year == 2010').sample(5)

game_id  year  month  day_ res_batter  H/PA    H   PA  AB   BB  \
38615  BOS201005110  2010      5    11   bautj002   0.0  0.0  3.0   3  0.0   
38671  BAL201007180  2010      7    18   bautj002   0.0  0.0  5.0   4  1.0   
38625  ARI201005220  2010      5    22   bautj002   0.0  0.0  3.0   3  0.0   
38598  TBA201004240  2010      4    24   bautj002   0.0  0.0  4.0   3  0.0   
38638  TOR201006060  2010      6     6   bautj002   0.0  0.0  4.0   3  1.0   

       ...  HBP  ITF   1B   2B   3B   HR  SH  SF        BA       OBP  
38615  ...  0.0  0.0  0.0  0.0  0.0  0.0   0   0  0.210938  0.317881  
38671  ...  0.0  0.0  0.0  0.0  0.0  0.0   0   0  0.233438  0.355263  
38625  ...  0.0  0.0  0.0  0.0  0.0  0.0   0   0  0.237500  0.343915  
38598  ...  1.0  0.0  0.0  0.0  0.0  0.0   0   0  0.227273  0.358025  
38638  ...  0.0  0.0  0.0  0.0  0.0  0.0   0   0  0.250000  0.377049  

[5 rows x 21 columns]

In [9]:
# We got what we want, we can now save the dataframe
batting_summary.to_csv('../2005-2018_game_logs.csv',index=None)

## Running Tests

We can run tests on the code and verify the results by comparing them to the information on a reputable website: baseball-reference.com

<br> To do this we will pick a year, here I've chosen 2006 but this can easily be changed

<br> Next we will randomly pick a batter and visit his website and manually check the numbers

In [10]:
yearly_batting_summary = batting_summary.groupby('year')

In [11]:
df2006 = yearly_batting_summary.get_group(2006)

### Case 1: Checking Hits, AB, and PA

In [15]:
import random
batter = random.sample(set(df2006.res_batter),1)
print(batter)

df2006[df2006.res_batter == batter[0]].H.sum(), df2006[df2006.res_batter == batter[0]].AB.sum(), df2006[df2006.res_batter == batter[0]].PA.sum()

['bradm001']


(97.0, 351, 405.0)

The chosen batter is 'petta001' which is Milton Bradley (https://www.retrosheet.org/boxesetc/B/Pbradm001.htm) Milton Bradley was actually a pitcher but the code picked up some At-Bats for him <br><br> We can check his 2006 stats by following this link: https://www.baseball-reference.com/players/gl.fcgi?id=bradlmi01&t=b&year=2006
<br><br>By scrolling to the bottom of the web-page, we can see that he had exactly 405 Plate appearances, 351 AB and 97 Hits which corresponds to the data we obtained

### Case 2: Checking BB, IBB and HBP,

In [16]:
df2006[df2006.res_batter == batter[0]].BB.sum(),df2006[df2006.res_batter == batter[0]].IBB.sum(),df2006[df2006.res_batter == batter[0]].HBP.sum()

(50.0, 1.0, 2.0)

Checking his Walks, Intentional Walks and Hit by Pitch on baseball-reference we can see that he has 51 Walks, with 1 intentional walks and 6 Hit by Pitches. <br><br>I've double checked the game date below and I can confirm that baseball-reference counts the Intentional walks in the Walks Columns which is why they have it at 51 (50 normal walks + 1 intentional walks) instead of 50

In [18]:
df2006.query('res_batter =="bradm001" and month==7 and day_ ==31')[['game_id','year','month','day_','res_batter','BB','IBB']]

game_id  year  month  day_ res_batter   BB  IBB
70781  ANA200607310  2006      7    31   bradm001  0.0  1.0

In [20]:
df.query('year==2006 and res_batter =="bradm001" and month==7 and day_ ==31')

game_id away_team  inning  batting_team  outs  balls  strikes  \
10133  ANA200607310       OAK       1             0     1      3        2   
10154  ANA200607310       OAK       4             0     0      3        0   
10171  ANA200607310       OAK       6             0     2      2        0   
10187  ANA200607310       OAK       8             0     1      3        0   

       vis_score  home_score res_batter  ... double_header_flag hit_flag   BB  \
10133          0           0   bradm001  ...                  0      0.0  0.0   
10154          0           0   bradm001  ...                  0      1.0  0.0   
10171          0           1   bradm001  ...                  0      1.0  0.0   
10187          3           1   bradm001  ...                  0      0.0  0.0   

       IBB  HBP  ITF   1B   2B   3B   HR  
10133  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
10154  0.0  0.0  0.0  1.0  0.0  0.0  0.0  
10171  0.0  0.0  0.0  0.0  0.0  0.0  1.0  
10187  1.0  0.0  0.0  0.0  0.0  0.0  0.0  

[4 rows x 50 columns]

Henry